# MMcPaper
Scripts for MeDIP-Seq analyses described in the paper "The vertical transfer of maternal immune cells promotes neonatal immunity against viral infections".

(g) Phylogenetic  distances  between  intestinal microbiome  samples,  calculated  via  weighted  UniFrac  distance  matrix.  Each  circle  in  the  scatter  plot  represents  a microbiome sample, collected on postnatal day 8 from both mother and neonates. The percentage of variation explained by each  principal  component  (PC) is  indicated  on  the  axes.  WT  (n=5)  and  Rag2-/-γc-/-(n=4)  C57BL/6  mothers  (F0)  have significantly different weighted UniFrac distance based on ANOSIM(P = 0.001) analysis while MMc+(n=16) and MMclow(n=13) offspring (F1) on postnatal day 8 do not significantly differfrom each other.Relative abundances at the phylum (h) and class (i) level within the taxonomic composition of the gut microbiome in WT and MMc+ and MMclowoffspring (both Rag2-/-γc-/-dams and their MMc+ and MMclowoffspring. Statistics on (g) are described in the methods part, and dots represent individual samples.

## METHOD MeDIP-Seq 
MeDIP  (Methylation  Dependent  ImmunoPrecipitation)  was  performed  as  recently  published,  enriched  methylated DNA regions were identified by short read sequencing. Briefly, HSC were sorted from bone marrow of 10-11 week old MMc+ and MMclow offspring, following the gating strategy depicted in Fig. S4.Genomic DNA was extracted from 671 1.5x105 HSC upon resuspending the cells in 300 μl lysis buffer (7,5% NaHCO3) by phenol-chloroform extraction and 672 EtOH precipitation. Isolated genomic DNA was fragmented by sonification (Bioruptor, Diagenode, 25 cycles (30s on, 673 30s  off)  to  an  average  fragment  size  of  50–100  bp.  For  quality  control,  0.025  ng  DNA  from  in  vitro  methylated  674 pCR2.1 plasmid were added to each sample as a spike-in control. DNA was denatured at 98°C for 10 min and cooled 675 on ice for 10 min. From each sample, 1/10 of the total amount was used an input control. For MeDIP pull-down 5 μg 676 of a 5-mc-specific antibody (Diagenode) were added to the rest of the sample and rotated for 2h at 4°C. Subsequently, 677 50 μl M-280 anti-mouse IgG Dynabeads were added and the complexes were incubated on a rotation wheel for 2h at 678 4°C. The following steps were performed in parallel with the input samples. RNA digestion was performed at 37°C 679 for 30 min using 1.3 μl RNAse A (20-40 mg/ml) prior to elution and protein digestion at 55°C for 30 min using 200 680 μl elution buffer (50mM Tris-HCl, 10 mM EDTA and 1 % SDS), 4 μl proteinase K (40 mg/ml) and 7 μl CaCl2 (300 681 mM). DNA was extracted using phenol-chloroform extraction and EtOH precipitation as described above. 682 MeDIP compatible next generation sequencing libraries were generated using 5 ng DNA and the DNA SMART ChIP-683 Seq  Kit  (TaKaRa,  Clontech)  suitable  for  single  stranded  DNA.  MeDIP  libraries  were  sequenced  on  a  NextSeq  500  684 system (Illumina) using single read (1x75) flow cells and 30 million reads per MeDIP sample and 10 million reads per 685 Input control. 686 After  quality  control  of  MeDIP-seq  data  which  is  performed  using  FASTQC75,  reads  were  aligned  to  the  MM10  687 genome using  BWA76,  resulting  in  BAM  files.  The  alignment  files  were  analyzed  using  the  IGV  software.  For  the  688 detection  of  differentially  methylated  regions  (DMRs)  and  gene  ontology  analyses,  the  tool  diffreps77  was  utilized.  689 Default  parameters  were  used  except  for  the  adaption  of  the  window  size  to  800  nt.  The  genes  with  significantly  690 DMRs (P value < 0.0001) were filtered and analyzed in IGV.

## METHOD Microbiome analysis 
V3–V4  region  16S  rRNA  amplicons  were  generated  using  degenerate  primers  that  contain  the  Illumina  adapter consensus sequence F: (5’TCGTCGGCAGCGTCAGATGTGT ATAAGAGACA-GCCTACGGGNGGCWGCAG-3’) and  R:  (5’GTCTCGTGGGCTCGGAG  ATGTGTATAAGAGACAGGACTACHVGGGTATCTAATCC-3`)81,82.  A detailed  description  of  the  protocol  has  been  published  online  by  Illumina:  ‘16S  Metagenomic  Sequencing  Library Preparation:  Preparing  16S  Ribosomal  RNA  Gene  Amplicons  for  the  Illumina  MiSeq  System’.  Samples  were multiplexed  using  the  Illumina  Nextera  XT  Index  Kit  to  construct  barcoded  libraries.  Subsequent  libraries  were multiplexed  and  sequenced  by  500PE  sequencing  on  the  MiSeq  platform.  FastQC  (Babraham  Bioinformatics, Babraham  Institute,  UK)  was  used  to  determine  the  average  quality  scores  of  each  sample  before  and  after  paired reads. The paired ends in each sample were joined, and all sequences less than 250 bp and/or with a Phred score <33 were discarded. Quality filtering was applied using QIIME 53 (at Phred ≥ Q20). We performed operational taxonomic unit (OTU) clustering and alpha- and beta-diversity analysis using QIIME version 1.7 83. A chimera filter was applied using  USEARCH  8.1.  All  sequences  were  clustered  based  on  97%  similarity  to  reference  sequences. The reads that did not meet the 97% similarity criteria were clustered de novo. Taxonomy levels of representative sequences in the OTU table were assigned at 95% similarity based on the SILVA database. We calculated alpha diversity based on the Shannon  diversity  index.  Analysis  of  beta  diversity  statistics  (analysis  of  similarities,  ANOSIM)  was  performed  to determine if differences between the distributions of microbiota profiles from the five datasets were significant. 

